### PR TITLE
[process-agent] Better process collection on Windows

### PR DESCRIPTION
**Pull request by xornivore**
### What does this PR do?

- Adds probe implementation based on using Windows PDH performance counter APIs which has has the following benefits over the existing `toolhelp` API collector:
  - No need to use process handles to collect process stats
  - Largely reduced number of API calls - we can collect data for all processes using very few calls to get PDH formatted value arrays.
  - New implementation can be turned on by setting `process_config.windows.use_perf_counters`, however the previous check implementation based on the `toolhelp` API will remain the default in the next release.

- Redefines `procutil.Probe` as an interface to allow for multiple implementations for the same OS.
- Consolidate `Process` and `RTProcess` checks:
    - Defines a new interface for a check that can run both regular and RT data collection (`CheckWithRealTime`)
    - Implements a custom check runner that performs skips of RT interval when RT is not enabled.
    - Consolidates functionality in `Process` and `RTProcess` allowing to use the same instance of `procutil.Probe`.
    - Eliminates the need for synchronization between regular and RT process data collection.
  
### Motivation

- Better, more maintainable process collection on Windows.
- Eliminate the need for `procutil.Probe` implementations to be thread safe and eliminate the synchronization between `RTProcess` and `Process` checks. 

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

These are pretty large changes that touch process collection code for multiple platforms.

- Validate that process and RT collection works correctly in all supported platforms.
- Validate the use of perf counter based probe and compare it to the existing legacy toolhelp implementation across supported versions of Windows.
- Validate that the `Connections` check can still properly use create time info from the `Process` check.

### Checklist

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: tasks/go.py [hunk 1]
```diff
@@ -21,14 +21,15 @@
 from .utils import get_build_flags
 
 # List of modules to ignore when running lint
-MODULE_WHITELIST = [
+MODULE_ALLOWLIST = [
```
Comment (mx-psi) [inline tasks/go.py:24]: Can you do this name change (and the change in the printed message) in a separate PR?

File: tasks/go.py [hunk 2]
```diff
@@ -21,14 +21,15 @@
...
     # Windows
     "doflare.go",
     "iostats_pdh_windows.go",
     "iostats_wmi_windows.go",
     "pdh.go",
     "pdh_amd64.go",
     "pdh_386.go",
+    "pdhformatter.go",
```
Comment (mx-psi) [inline tasks/go.py:32]: Why is this needed?

File: tasks/go.py (hunk 1 above)
Comment (xornivore) [reply tasks/go.py:24]: Sounds good, I will extract this into a separate PR.

File: tasks/go.py (hunk 2 above)
Comment (xornivore) [reply tasks/go.py:32]: So I was able to work around the issue causing the warning about `misuse of unsafe.Pointer`. At first I misinterpreted as belonging to the lint output (hence the changes to this invoke task), but in fact it is from `go vet` and there isn't a good way to mark exclusions for it, `// go vet: .. ` does seem to not work. Afaict the only option as of today is to build with `!dovet`. To summarize - once I confirm I should be able to remove addition of this file to the allow list.

Comment (mx-psi) [reply tasks/go.py:32]: Yeah, `go vet` doesn't have an exclusion mechanism sadly :( we can use the skip list if you don't find another way around it

File: pkg/process/procutil/process_windows.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+	proc.Stats.CreateTime = ctime
+	return nil
+}
+
+// TODO: deduplicate
+func openProcHandle(pid int32) (windows.Handle, error) {
+	// 0x1000 is PROCESS_QUERY_LIMITED_INFORMATION, but that constant isn't
+	//        defined in x/sys/windows
+	// 0x10   is PROCESS_VM_READ
+	procHandle, err := windows.OpenProcess(0x1010, false, uint32(pid))
```
Comment (just-chillin) [inline pkg/process/procutil/process_windows.go:440]: nit: Can we create a constant for 0x1010 and 0x10?

File: pkg/util/winutil/pdhutil/pdh.go [hunk 4]
```diff
@@ -132,6 +132,42 @@ const (
...
+	CounterAllProcessIODataOpsPerSec    = `\Process(*)\IO Data Operations/sec`
+	CounterAllProcessIOOtherOpsPerSec   = `\Process(*)\IO Other Operations/sec`
+	CounterAllProcessIOReadBytesPerSec  = `\Process(*)\IO Read Bytes/sec`
+	CounterAllProcessIOWriteBytesPerSec = `\Process(*)\IO Write Bytes/sec`
+	CounterAllProcessIODataBytesPerSec  = `\Process(*)\IO Data Bytes/sec`
+	CounterAllProcessIOOtherBytesPerSec = `\Process(*)\IO Other Bytes/sec`
+	CounterAllProcessWorkingSetPrivate  = `\Process(*)\Working Set - Private`
+)
+
+type PDH_FMT_COUNTERVALUE_ITEM struct {
```
Comment (just-chillin) [inline pkg/util/winutil/pdhutil/pdh.go:166]: Can we add a comment explaining what this does, and linking it to the microsoft docs for the one in `pdh.h`?

https://docs.microsoft.com/en-us/windows/win32/api/pdh/ns-pdh-pdh_fmt_countervalue_item_a

File: pkg/process/procutil/process_linux.go [hunk 5]
```diff
@@ -53,34 +53,37 @@ type statInfo struct {
...
+		if linuxProbe, ok := p.(*probe); ok {
+			linuxProbe.bootTimeRefreshInterval = bootTimeRefreshInterval
+		}
 	}
 }
 
-// Probe is a service that fetches process related info on current host
-type Probe struct {
+// probe is a service that fetches process related info on current host
+type probe struct {
```
Comment (just-chillin) [inline pkg/process/procutil/process_linux.go:86]: I think we should probably rename this to something more meaningful.

File: pkg/util/winutil/pdhutil/pdhformatter.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		uintptr(unsafe.Pointer(&bufLen)),
+		uintptr(unsafe.Pointer(&itemCount)),
+		uintptr(unsafe.Pointer(&buf[0])),
+	)
+	if r != ERROR_SUCCESS {
+		return fmt.Errorf("Error getting formatted counter array 0x%x", r)
+	}
+
+	var items []PDH_FMT_COUNTERVALUE_ITEM
+	hdrItems := (*reflect.SliceHeader)(unsafe.Pointer(&items))
```
Comment (just-chillin) [inline pkg/util/winutil/pdhutil/pdhformatter.go:68]: The inner rust developer in me sets off alarms whenever I see an unsafe pointer dereference. Is it possible for any part of this to fail?

I would add a comment explaining why this is safe to do, and how we know that we won't segfault whenever you use unsafe.

File: pkg/process/checks/allprocesses_windows.go [hunk 7]
```diff
@@ -28,6 +29,7 @@ var (
 
 	// XXX: Cross-check state is stored globally so the checks are not thread-safe.
 	cachedProcesses = map[uint32]*cachedProcess{}
+	// TODO: remove locking here
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/process/checks/allprocesses_windows.go:32]: With these changes we no longer have to guard cachedProcesses with a lock because we are using the same goroutine to run both regular and RT checks (there is no contention).

File: pkg/util/winutil/pdhutil/pdhformatter.go (hunk 6 above)
Comment (xornivore) [reply pkg/util/winutil/pdhutil/pdhformatter.go:68]: Adding a comment sounds like a good idea.

File: pkg/process/procutil/process_linux.go (hunk 5 above)
Comment (xornivore) [reply pkg/process/procutil/process_linux.go:86]: For the sake of keeping this PR smaller probably worth splitting that, though I think pondering a good alternative might take time - collector (taken), enumerator (odd), if you have a suggestion feel free to post here.

File: pkg/util/winutil/pdhutil/pdh.go (hunk 4 above)
Comment (xornivore) [reply pkg/util/winutil/pdhutil/pdh.go:166]: I think generally this is clear we are dealing with PDH APIs, but adding a comment to the struct makes sense.

File: pkg/process/procutil/process_windows.go (hunk 3 above)
Comment (xornivore) [reply pkg/process/procutil/process_windows.go:440]: yep, this is duplicated in a couple of places, probably best done in a separate PR, it belongs in the `pkg/util/winutil`.

File: pkg/process/checks/process.go [hunk 8]
```diff
@@ -33,9 +34,7 @@ type ctrProcMsgFactory func([]*model.Process, ...*model.Container) *model.Collec
 // for live and running processes. The instance will store some state between
 // checks that will be used for rates, cpu calculations, etc.
 type ProcessCheck struct {
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/process/checks/process.go:36]: I suspect rethinking that singleton pattern is a separate effort - they are in the "all checks" slice, and many other checks would need to change.

File: pkg/process/checks/process.go [hunk 9]
```diff
@@ -22,7 +21,9 @@ import (
 const emptyCtrID = ""
 
 // Process is a singleton ProcessCheck.
-var Process = &ProcessCheck{probe: procutil.NewProcessProbe()}
+var Process = &ProcessCheck{}
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/process/checks/process.go:24]: Agreed, in principle, in fact each check already implements the exported `Check` interface and these are currently consumed in [cmd/process-agent](https://github.com/DataDog/datadog-agent/blob/main/cmd/process-agent/collector.go#L74) via the [All](https://github.com/DataDog/datadog-agent/blob/main/pkg/process/checks/checks.go#L20) slice of `Check` checks. If we decide to change the fact that the names are exported, it's probably a change better consolidated to a separate PR because we'd want to rename other checks as well.

File: pkg/process/checks/process.go [hunk 10]
```diff
@@ -45,16 +44,26 @@ type ProcessCheck struct {
...
+	// will be reused by RT process collection to get stats
+	lastPIDs []int32
+
+	// Create times by PID used in the network check
+	createTimes atomic.Value
 }
 
 // Init initializes the singleton ProcessCheck.
-func (p *ProcessCheck) Init(_ *config.AgentConfig, info *model.SystemInfo) {
+func (p *ProcessCheck) Init(cfg *config.AgentConfig, info *model.SystemInfo) {
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/process/checks/process.go:63]: Do you mean an equivalent of `NewProcessCheck` where we create and initialize it?

File: cmd/process-agent/main_common.go [hunk 11]
```diff
@@ -239,40 +240,79 @@ func debugCheckResults(cfg *config.AgentConfig, check string) error {
...
+
+	printResultsBanner(ch.RealTimeName())
+
+	runs, err := ch.RunWithOptions(cfg, 1, options)
+	if err != nil {
+		return fmt.Errorf("collection error: %s", err)
+	}
+
+	for _, r := range runs {
+		if r.CheckName == ch.RealTimeName() {
```
Comment (shang-wang) [inline cmd/process-agent/main_common.go:302]: nit: this can be removed.

File: pkg/process/checks/process.go [hunk 12]
```diff
@@ -110,14 +131,13 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
 	}
 
 	// stores lastPIDs to be used by RTProcess
-	lastPIDs := make([]int32, 0, len(procs))
+	p.lastPIDs = make([]int32, 0, len(procs))
```
Comment (shang-wang) [inline pkg/process/checks/process.go:134]: nit: since everything is synchronous, we could use `p.lastPIDs = p.lastPIDs[:0]` to avoid allocating new memory.

File: pkg/process/checks/process.go [hunk 13]
```diff
@@ -78,9 +90,17 @@ func (p *ProcessCheck) RealTime() bool { return false }
...
-	p.Lock()
-	defer p.Unlock()
+	results, err := p.run(cfg, groupID, false)
+	if err != nil {
+		return nil, err
+	}
+	if len(results) == 0 {
+		return nil, nil
+	}
+	return results[0].Messages, nil
```
Comment (shang-wang) [inline pkg/process/checks/process.go:100]: Sorry I was a bit confused by this, sounds like the `results` is only going to have either one or two items depend on whether it's RT or not? If so why not creating a struct? I have not reviewed the whole thing so I could be missing something.

File: pkg/process/checks/runner.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+	ExitChan       chan struct{}
+	RtIntervalChan chan time.Duration
+	RtEnabled      func() bool
+	RunCheck       func(options RunOptions)
+}
+
+type runnerWithRealTime struct {
+	RunnerConfig
+	ratio      int
+	counter    int
```
Comment (shang-wang) [inline pkg/process/checks/runner.go:24]: Can we have some comment on how they are used?

File: pkg/process/checks/runner.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+	newTicker  func(d time.Duration) *time.Ticker
+	stopTicker func(t *time.Ticker)
```
Comment (shang-wang) [inline pkg/process/checks/runner.go:26]: nit: why can't the `newTicker` be a `ticker` property of `runnerWithRealTime`? And why `stopTicker` has to be a property instead?

File: pkg/process/config/yaml_config.go [hunk 16]
```diff
@@ -187,6 +187,11 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 		a.Windows.AddNewArgs = config.Datadog.GetBool(addArgsKey)
 	}
 
+	// Windows: Controls using the new check based on performance counters PDH APIs
+	if usePerfCountersKey := key(ns, "windows", "use_perf_counters"); config.Datadog.IsSet(usePerfCountersKey) {
```
Comment (shang-wang) [inline pkg/process/config/yaml_config.go:191]: nit: the config has to be added as `SetKnown()` in `pkg/config/config.go` otherwise it'll report warning when agent starts and parsing this config.

Comment (xornivore) [reply pkg/process/config/yaml_config.go:191]: That's a good point, I forgot about it.

File: pkg/process/checks/runner.go (hunk 15 above)
Comment (xornivore) [reply pkg/process/checks/runner.go:26]: Practically it's just dependency injection for unit tests, `ticker` is re-created on RT change in the goroutine, and stopping a ticker panics if it's not initialized (in cases when we inject it, it's not properly initialized, just the channel is initialized.

File: pkg/process/checks/process.go (hunk 13 above)
Comment (xornivore) [reply pkg/process/checks/process.go:100]: I agree with you, it's quite clunky as is in the end, I will change results to be a struct with two fields.

File: pkg/process/checks/allprocesses_windows.go [hunk 17]
```diff
@@ -25,32 +25,8 @@ var (
...
-	cachedProcesses = map[uint32]*cachedProcess{}
-	// cacheProcessesMutex is a mutex to protect cachedProcesses from being accessed concurrently.
-	// So far this is the case for Process check and RTProcess check
-	// TODO: revisit cacheProcesses usage so that we don't need to lock the whole getAllProcesses()
-	cacheProcessesMutex sync.Mutex
-	checkCount          = 0
-	haveWarnedNoArgs    = false
 )
 
-type SystemProcessInformation struct {
```
Comment (xornivore) [inline pkg/process/checks/allprocesses_windows.go:39]: Was not used here

**Review by ISauve — APPROVED** [review]
very shallow review as I only looked at the file owned by the `network-agent` team - but LGTM!

**Review by djmitche — APPROVED** [review]
Also a shallow review of the winutil bits. The PR description sounds good, especially the use of a feature gate.

File: pkg/process/checks/process.go [hunk 18]
```diff
@@ -133,7 +151,15 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
...
 		p.lastCtrIDForPID = ctrByProc
 		p.lastRun = time.Now()
-		return nil, nil
+		p.storeCreateTimes()
+
+		if collectRealTime {
+			p.realtimeLastCPUTime = p.lastCPUTime
+			p.realtimeLastProcs = procsToStats(p.lastProcs)
+			p.realtimeLastCtrRates = p.lastCtrRates
+			p.realtimeLastRun = p.lastRun
```
Comment (wiyu) [inline pkg/process/checks/process.go:160]: nit: Would be nice if the storing of the last checks was in a function, but maybe in a future PR.

File: pkg/process/checks/process.go [hunk 19]
```diff
@@ -316,17 +393,29 @@ func fmtProcesses(
 		// Hide blacklisted args if the Scrubber is enabled
 		fp.Cmdline = cfg.Scrubber.ScrubProcessCommand(fp)
 
+		var ioStat *model.IOStat
+		if fp.Stats.IORateStat != nil {
```
Comment (wiyu) [inline pkg/process/checks/process.go:397]: What's the difference btw this `IORateStat` and the `IOStat`?

File: pkg/process/checks/process.go (hunk 18 above)
Comment (xornivore) [reply pkg/process/checks/process.go:160]: Agreed, I will do some refactoring in a separate PR along with more tests, now that the probe is a mockable interface it should be fairly easy to do that.

File: pkg/process/checks/process.go (hunk 19 above)
Comment (xornivore) [reply pkg/process/checks/process.go:397]: I had to abstract the rate (in /sec measurements) vs increasing values (that we convert to rates):
```
// IOCountersStat holds IO metrics for a process
type IOCountersStat struct {
	ReadCount  int64
	WriteCount int64
	ReadBytes  int64
	WriteBytes int64
}

// IOCountersRateStat holds IO metrics for a process represented as rates (/sec)
type IOCountersRateStat struct {
	ReadRate       float64
	WriteRate      float64
	ReadBytesRate  float64
	WriteBytesRate float64
}
```

File: cmd/process-agent/collector.go [hunk 20]
```diff
@@ -105,9 +105,43 @@ func (l *Collector) runCheck(c checks.Check, results *api.WeightedQueue) {
...
-		log.Errorf("Unable to run check '%s': %s", c.Name(), err)
+
+	if withRealTime, ok := c.(checks.CheckWithRealTime); ok {
+		run, err := withRealTime.RunWithOptions(l.cfg, l.nextGroupID, options)
+		if err != nil {
+			log.Errorf("Unable to run check '%s': %s", c.Name(), err)
+			return
+		}
+		l.messagesToResults(start, withRealTime.Name(), run.Standard, results)
+		l.messagesToResults(start, withRealTime.RealTimeName(), run.RealTime, results)
```
Comment (mbotarro) [inline cmd/process-agent/collector.go:116]: Is the same `api.WeightedQueue` passed to both regular and RT checks? https://github.com/DataDog/datadog-agent/pull/8875 created a separate queue for RT checks. I might be missing something but it seems that we're using the same queue for both checks now.

File: cmd/process-agent/collector.go [hunk 21]
```diff
@@ -105,9 +105,43 @@ func (l *Collector) runCheck(c checks.Check, results *api.WeightedQueue) {
...
+			return
+		}
+		l.messagesToResults(start, c.Name(), messages, results)
+	}
+
+	if !c.RealTime() || options.RunStandard {
+		d := time.Since(start)
+		switch {
+		case runCounter < 5:
+			log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
```
Comment (mbotarro) [inline cmd/process-agent/collector.go:130]: should we only increment `runCounter` when the normal check is run?

File: cmd/process-agent/collector.go (hunk 20 above)
Comment (xornivore) [reply cmd/process-agent/collector.go:116]: Nice catch, yep, this PR predates #8875, I will fix this.

File: cmd/process-agent/collector.go (hunk 21 above)
Comment (xornivore) [reply cmd/process-agent/collector.go:130]: We do check for it [here](https://github.com/DataDog/datadog-agent/pull/8846/files/08c387d53948737801e468cd267263f3b15968e2#diff-993d2491e7efac8e13349ef7859c375a68e328aa5415289b699a6da0cc2637c3R126)

File: pkg/process/checks/process.go (hunk 18 above)
Comment (shang-wang) [reply pkg/process/checks/process.go:160]: Even better we could wrap the `last***` and `realtimeLast***` variables into structs.

Comment (xornivore) [reply pkg/process/checks/process.go:160]: yep, that's a great idea and it would prevent obvious typos in some cases

File: cmd/process-agent/collector.go (hunk 21 above)
Comment (mbotarro) [reply cmd/process-agent/collector.go:130]: If `runCheck` is called for both standard and RT checks, won't `runCounter` be increased by RT checks? 
https://github.com/DataDog/datadog-agent/blob/08c387d53948737801e468cd267263f3b15968e2/cmd/process-agent/collector.go#L99-L103

Comment (xornivore) [reply cmd/process-agent/collector.go:130]: yes, but RT check in that case is an illusion because we are just copying stats from the standard check

File: cmd/process-agent/collector.go (hunk 20 above)
Comment (xornivore) [reply cmd/process-agent/collector.go]: done.

File: pkg/process/procutil/process_model.go [hunk 22]
```diff
@@ -33,13 +62,55 @@ type Stats struct {
...
+		copy.IOStat = &IOCountersStat{}
+		*copy.IOStat = *s.IOStat
+	}
+	if s.IORateStat != nil {
+		copy.IORateStat = &IOCountersRateStat{}
+		*copy.IORateStat = *s.IORateStat
+	}
+	if s.CtxSwitches != nil {
+		copy.CtxSwitches = &NumCtxSwitchesStat{}
+		*copy.CtxSwitches = *s.CtxSwitches
```
Comment (just-chillin) [inline pkg/process/procutil/process_model.go:109]: Why are we writing to the struct twice?

Comment (xornivore) [reply pkg/process/procutil/process_model.go:109]: we are initializing the pointer, then copying the data to the struct.

**Review by just-chillin — APPROVED** [review]
LGTM

File: cmd/process-agent/main_common.go [hunk 23]
```diff
@@ -320,40 +321,82 @@ func debugCheckResults(cfg *config.AgentConfig, check string) error {
...
+		groupID     int32
+		nextGroupID = func() int32 {
+			groupID++
+			return groupID
+		}
+	)
+
+	if _, err := ch.RunWithOptions(cfg, nextGroupID, options); err != nil {
+		return fmt.Errorf("collection error: %s", err)
+	}
```
Comment (mbotarro) [inline cmd/process-agent/main_common.go:379]: nit: should we add a comment to explain why we need to run the check twice?

Comment (xornivore) [reply cmd/process-agent/main_common.go:379]: good point I will add that 🙇‍♂️

File: pkg/process/procutil/process_model.go [hunk 24]
```diff
@@ -94,6 +171,14 @@ type IOCountersStat struct {
 	WriteBytes int64
 }
 
+// IOCountersRateStat holds IO metrics for a process represented as rates (/sec)
```
Comment (shang-wang) [inline pkg/process/procutil/process_model.go:174]: nit: let's not insert this definition between `IOCountersStats` and `func (i *IOCountersStat) IsZeroValue()` which ideally should be together.

Comment (xornivore) [reply pkg/process/procutil/process_model.go:174]: Makes sense, it's an oversight on my part.

File: pkg/process/checks/runner.go [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+			}
+		}
+	}
+}
+
+func getRtRatio(checkInterval, rtInterval time.Duration) (int, error) {
+	if checkInterval <= rtInterval {
+		return -1, errors.New("check interval should be larger than RT interval")
+	}
+	if checkInterval%rtInterval != 0 {
```
Comment (shang-wang) [inline pkg/process/checks/runner.go:95]: Just want to make sure that this is defensive programming right? Because I was checking the config section, and the logic would use default intervals if the config doesn't make sense.

File: pkg/process/config/yaml_config.go [hunk 26]
```diff
@@ -94,6 +94,19 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 	a.setCheckInterval(ns, "process_realtime", RTProcessCheckName)
 	a.setCheckInterval(ns, "connections", ConnectionsCheckName)
 
+	if a.CheckIntervals[ProcessCheckName]%a.CheckIntervals[RTProcessCheckName] != 0 {
```
Comment (shang-wang) [inline pkg/process/config/yaml_config.go:97]: Should there be code to also make sure regular check interval is always greater than RT check?

File: pkg/process/checks/runner.go (hunk 25 above)
Comment (xornivore) [reply pkg/process/checks/runner.go:95]: yes, it is defensive coding

File: pkg/process/config/yaml_config.go (hunk 26 above)
Comment (xornivore) [reply pkg/process/config/yaml_config.go:97]: Great point I will add that.

File: pkg/process/checks/process.go (hunk 19 above)
Comment (shang-wang) [reply pkg/process/checks/process.go:397]: Took a minute for me to realized that `IOCountersRateStat` is used in Windows and the old `IOCountersStat` is used in others(am I right?). If it's the case I feel like explicitly calling them out as comments on the model file would make it easier to understand.

File: cmd/process-agent/collector.go [hunk 27]
```diff
@@ -315,6 +329,71 @@ func (l *Collector) run(exit chan struct{}) error {
...
+				RunCheck: func(options checks.RunOptions) {
+					l.runCheckWithRealTime(withRealTime, results, rtResults, options)
+				},
+			},
+		)
+	}
+	return l.basicRunner(c, results, exit), nil
+}
+
+func (l *Collector) basicRunner(c checks.Check, results *api.WeightedQueue, exit chan struct{}) func() {
```
Comment (shang-wang) [inline cmd/process-agent/collector.go:367]: `legacyRunner`?

Comment (xornivore) [reply cmd/process-agent/collector.go:367]: It's not really legacy, maybe `simpleRunner`? because there are checks that will always be using this (e.g. there's no realtime`pod` or `connections`).

Comment (shang-wang) [reply cmd/process-agent/collector.go:367]: Ah you are right, I missed the non-process checks.

Comment (xornivore) [reply cmd/process-agent/collector.go:367]: No problem, it would be nice to move some of this collector code under `pkg/process` including the runner code here. It's a bit busy.

File: pkg/process/procutil/process_windows.go [hunk 28]
```diff
@@ -0,0 +1,609 @@
...
+			log.Debugf("Couldn't open process %v %v", pid, err)
+			return windows.Handle(0), err
+		}
+	}
+	return procHandle, nil
+}
+
+// GetUsernameForProcess returns username for a process
+func GetUsernameForProcess(h windows.Handle) (name string, err error) {
+	name = ""
```
Comment (shang-wang) [inline pkg/process/procutil/process_windows.go:549]: nit: these are not needed.

**Review by shang-wang — APPROVED** [review]
🔥

**Review by mx-psi — APPROVED** [review]
